lib/kojo/refinements/string.rb

```diff
@@ -46,6 +46,12 @@ def eval_erb(args, filename)
         raise Kojo::TemplateError, "#{e.message}\nin: #{filename}"
       end
 
+      def compress_imports
+        gsub /^ *@import [^(\s]*\s*\([^)]+\)$/ do |match|
```

Comment:
What do you think about these suggestions:

1. Adding `+` to cover the case when there is more than one space after `@import`:

```diff
- /^ *@import [^(\s]*\s*\([^)]+\)$/
+ /^ *@import +[^(\s]*\s*\([^)]+\)$/
```

2. Since `[^(\s]*` means imported file name, I'd propose replacing `*` with `+` because name cannot be empty: `[^(\s]+`

```diff
- /^ *@import +[^(\s]*\s*\([^)]+\)$/
+ /^ *@import +[^(\s]+\s*\([^)]+\)$/
```

3. There might be a case with trailing whitespace after final closing `)`:

```diff
- /^ *@import +[^(\s]+\s*\([^)]+\)$/
+ /^ *@import +[^(\s]+\s*\([^)]+\)\s*$/
```

4. We probably need to strip comments because they will mess up the resulting line:

```yaml
@import bla( # comment1
  arg1: "" # comment2
) # comment3
```

and after substitution:

```yaml
@import bla( # comment1 arg1: "" # comment2 ) # comment3
```

5. There is also a case when one of the arguments contains `)`:

```
@import bla(
  arg1: ")"
)
```

To cover it let's consider the argument list to be any character and do it non-greedily like this:

```diff
- /^ *@import +[^(\s]+\s*\([^)]+\)\s*$/
+ /^ *@import +[^(\s]+\s*\([\S\s]*?\)\s*$/
```

Reply:
Nice.

In regards to comments, I think they should stay out of scope. Inline comments in multiline import should not be supported.

The other suggestions work fine, except the ending space matching, which "swallows" the newline after the closing brace. This was caught by the specs. I updated your regex to this:

```diff
- /^ *@import +[^(\s]+\s*\([\S\s]*?\)\s*$/
+ /^ *@import +[^(\s]+\s*\([\S\s]*?\) *$/
```

which puts the specs back in green.

Reply:
Hm, now I think it has even more potential issues with that grammar: removing trailing `# comments` with regex will require a pretty complex regex to distinguish those cases:

```yaml
@import bla(
  arg1: "# not a comment",
  arg2: "some value" # comment
  arg3: "some value" # comment that should not be treated as trailing ) even if it has at the end this symbol: )
)
```

It's possible to mitigate by using explicit directive like `@endimport`:

```yaml
@import bla
  arg1: "# not a comment",
  arg2: "some value" # comment
  arg3: "some value" # comment that should not be treated as trailing ) even if it has at the end this symbol: )
@endimport
```

But this `@endimport` goes against the YAML readability principles, in my opinion.


Also, strictly speaking, strings in YAML do not need to be specified in quotes (https://stackoverflow.com/a/69850618/2867273). With that, those cases

```yaml
@import bla(
  arg1: value()
)

@import bla(
  arg1: "value()"
)
```

should be identical, but with regex, we can't do that while keeping the regex reasonable.

With all of that, I tend to think now that to implement multiline `@import` we need to evaluate them not before parsing YAML (like it is implemented now), but after that. Essentially, evaluate all `@import` nodes in the YAML tree and replace them with the "imported" node from another file.

```yaml
--- parent.yaml
@import child:
  arg1: "# not a comment",
  arg2: "some value" # comment

--- child.yaml
some_real_stuff:
  - some how utilize ${arg1} those arguments ${arg2}
```

It looks much more complex than just replacing "\n" with a regex 🙂

Reply:
In response to https://github.com/DannyBen/kojo/pull/41#discussion_r812895215 :

Yeah, I agree that implementing "full syntax" support like in the comment above will require a lot of work, it makes sense to keep it how it is now with a simple regex.

Reply:
No no. Kojo is not YAML specific. It is format-agnostic.

The arguments list inside the `( ... )` of the import are not YAML, they are Ruby hash.

inline comments are out of scope, as well as `")"` as an argument.

Reply:
Do you want to test this on your real project, to see that it a) works as advertised and b) doesn't break other things?

Reply:
Yup, it looks great for my cases, thanks!